src/clj/puppetlabs/http/client/async.clj

```diff
@@ -117,11 +124,68 @@
       (.setAs (clojure-response-body-type->java opts))
       (.setBody (:body opts))
       (.setDecompressBody (clojure.core/get opts :decompress-body true))
-      (.setHeaders (:headers opts))))
+      (.setHeaders (:headers opts))
+      (.setMetricId (parse-metric-id opts))))
+
+(schema/defn get-mean :- schema/Num
+  [timer :- Timer]
+  (->> timer
+       .getSnapshot
+       .getMean
+       (.toMillis TimeUnit/NANOSECONDS)))
+
+(defn get-metric-data
+  [timer metric-id]
+  (let [count (.getCount timer)
+        mean (get-mean timer)
+        aggregate (* count mean)]
+    {:count count
+     :mean mean
+     :aggregate aggregate
+     :metric-id metric-id}))
+
+(defn get-metrics-data
+  [timers]
+  (reduce (fn [acc [metric-id timer]]
+            (assoc acc metric-id (get-metric-data timer metric-id)))
+          {} timers))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Public
 
+(schema/defn ^:always-validate get-client-metrics :- (schema/maybe common/Metrics)
+  "Returns the http client-specific metrics from the metric registry."
+  ([metric-registry :- common/OptionalMetricRegistry]
+   (when metric-registry
+     (into {} (Metrics/getClientMetrics metric-registry))))
+  ([metric-registry :- common/OptionalMetricRegistry
+    metric-filter :- common/MetricFilter]
+   (when metric-registry
+     (cond
+       (:verb metric-filter) (into {} (Metrics/getClientMetricsWithUrlAndVerb
+                                       metric-registry
+                                       (:url metric-filter)
+                                       (:verb metric-filter)
+                                       Metrics$MetricType/BYTES_READ))
+       (:url metric-filter) (into {} (Metrics/getClientMetricsWithUrl
+                                      metric-registry
+                                      (:url metric-filter)
+                                      Metrics$MetricType/BYTES_READ))
+       (:metric-id metric-filter) (into {} (Metrics/getClientMetricsWithMetricId
+                                            metric-registry
+                                            (into-array (:metric-id metric-filter)) Metrics$MetricType/BYTES_READ))
+       :else (throw (IllegalArgumentException. "Not an allowed metric filter."))))))
+
+(schema/defn ^:always-validate get-client-metrics-data :- (schema/maybe common/MetricsData)
+  "Returns a map of metric-id to metric data summary."
+  ([metric-registry :- common/OptionalMetricRegistry]
+   (let [timers (get-client-metrics metric-registry)]
+     (get-metrics-data timers)))
+  ([metric-registry :- common/OptionalMetricRegistry
+    metric-filter :- common/MetricFilter]
```

Comment:
> Sure? I guess I don't feel like it's that cumbersome/makes that much of a difference - {:metric-id [:foo] :metric-type :bytes-read} vs (metric-id-filter [:foo] :bytes-read), {:url "http://tickets.puppetlabs.com :metric-type :bytes-read} vs (url-filter "https://tickets.puppetlabs.com" :bytes-read), and{:url "http://tickets.puppetlabs.com :method "GET" :metric-type :bytes-read} vs (url-fitler "https://tickets.puppetlabs.com" "GET" :bytes-read)? Or we have (bytes-read-filter ...) and (response-init-filter ...) that handle adding in the metric-type? This seems like something that can be added in later without being a breaking change, so long as we keep the schema the same.

What I was getting at is that I think we should have a default value that we use for the `:bytes-read` part, because I don't think users are probably going to want to have to care about that in the majority of cases.  So the value in what I'm suggesting would be that we'd be giving them a way to build a filter without having to express that part.  It could be an optional argument to one of these functions, but in most cases hopefully they wouldn't need it.

I also think that that schema is pretty likely to evolve over time, so if we can steer things in a direction where the "public API" is to call a filter-creation function, e.g. `(metrid-id-filter ...)`, then we're only committed to maintaining the integrity of the function signature(s) as API, and we can move the schema around as much as we need in the future.

> Okay, I think based on going down this train of thought I am against standardizing the url in any way before making it a part of a metric name - that's why we have the metric id feature.

I think I'm good with that, just wanted to think it through, and consider codifying it in docs or something.

> I actually didn't realize that query strings were included in this. I think that they should be stripped out - I agree that they'll make for too many too specific timers, making it hard to aggregate.

I think that sounds best to me too.  I still think that regex support on the URL filter would be valuable but that definitely seems like something that could be additive API in the future.  Might be worth capturing in Jira if we all agree that it would be useful, though.

> I think I had asked about this in hipchat, and I think in that discussion we had decided to do just the exact match. But I think a reasonable API would be something like ["puppetlabs" "*"] if you want puppetlabs and everything under it, and ["puppetlabs"] if you just want the puppetlabs metric

Yeah, that may very well be the case.  My memory is terrible.  :)  This also seems like something we could probably add in the future without breaking API, right?  If so then maybe a Jira for that is sufficient for now.